orderer/consensus/follower/follower_chain.go

```diff
@@ -16,31 +18,49 @@ import (
 	"github.com/hyperledger/fabric/orderer/common/cluster"
 	"github.com/hyperledger/fabric/orderer/common/types"
 	"github.com/hyperledger/fabric/orderer/consensus"
+	"github.com/hyperledger/fabric/protoutil"
 	"github.com/pkg/errors"
 )
 
 //TODO skeleton
 
-//go:generate mockery -dir . -name Support -case underscore -output mocks
+//go:generate counterfeiter -o mocks/support.go -fake-name Support . Support
 
 // Support defines the interfaces needed by the follower.Chain, out of the much wider consensus.ConsenterSupport.
 type Support interface {
 	consensus.ConsenterSupport // TODO prune what we don't need, keep only what we do need.
 }
 
+const (
+	defaultPullRetryMinInterval time.Duration = 10 * time.Millisecond
+	defaultPullRetryMaxInterval time.Duration = 30 * time.Second
+)
+
 // Options contains all the configurations relevant to the chain.
 type Options struct {
-	Logger *flogging.FabricLogger
-	Cert   []byte
+	Logger               *flogging.FabricLogger
+	PullRetryMinInterval time.Duration
+	PullRetryMaxInterval time.Duration
+	Cert                 []byte
 }
 
 // CreateBlockPuller is a function to create BlockPuller on demand.
 // It is passed into chain initializer so that tests could mock this.
 type CreateBlockPuller func() (ChannelPuller, error)
 
+//go:generate counterfeiter -o mocks/puller_creator.go -fake-name BlockPullerCreator . BlockPullerCreator
+type BlockPullerCreator interface {
+	CreateBlockPuller() (ChannelPuller, error)
```

Comment:
Somewhat related to the comment I added in #1536, but it seems like we're introducing additional indirection in order to escape the consensus plugin box, and access resources in the rest of the orderer codebase that would not otherwise be accessible inside of a consensus plugin.

In general, I'd love to see this sort of indirection removed, have the block pulling resources created directly as real types, with a clear lifecycle and dependencies.  My expectation would be that the bulk of the logic in this PR could be preserved, but if we pulled it out of a consensus plugin, we could make a lot of the wiring much simpler -- not only for the new follower logic, but also for the inactive logic.

Reply:
Indeed following a channel is largely independent of consensus type, in the same way that the peer pulls blocks from endpoints. The dependency comes in when a follower needs to decide whether it is a member of the cluster. That is where the particular consensus implementation is needed. The cluster package solves that problem by indirectly receiving at creation a function pointer to a "AmIInChannel" method that is actually implemented by the etcdraft package. I also agree that casting the follower (and the inactive.Chain) as a chain is pretty artificial as it does not implement the important methods of the interface (Order, Configure). 

An alternative to the current composition would be to have a "follower" as a separate entity managed by the registrar, something that is not a consensus.Chain. It would reside in a package at the top level, like multichannel, etc. We would then have two maps in the registrar - chains and followers, mutually exclusive. A follower would know how to trigger the creation of a chain, and vise versa. Does that capture your intention?

Reply:
Yes, I think this is pretty close.  I would think we basically want to group our chains into three buckets.

1. Chains that this orderer is a member of the consenter set and is actively consenting on.
2. Chains that this orderer is not a member of the consenter set, but is tracking but might become a member of in the future. (inactive)
3. Chains that this orderer is not a member of the consenter set, but is actively replicating. (follower)

So, we will need a way for a consensus plugin to report, based on channel config, whether it is a member of the consenter set (the cluster service already does this via a function pointer, you say), but we could directly add it to the consenter interface if we wanted.

We will also need a way for an active consensus plugin to return control to the multichannel code indicating "I'm no longer active" and the multichannel code will need to decide whether the chain becomes inactive, or follower.  (I believe based on the RFC, we want this to be follower now?  Or perhaps that is only for the system-channel-less case?)

I would also think, that this could be a good place to fix the problem associated with state transfer and node removal, where the node is removed from the consenter set, and re-added say while offline.  If we coordinate the block puller, it could detect the removal, and stop the consensus plugin before completing the replication and restarting it.

Reply:
Note that the "I am no longer active" part is implemented by the etcdraft.Chain calling a "chain-creation" call-back, that is function pointer to the multichannel registrar "CreateChain" method. The same happens in the soon-to-be follower (when it wants to activate the etcdraft.Chain)  and the inactive-registry. So that is not a big issue.

Note that the "follower" bucket and the "inactive" bucket are not active at the same time, as we do not allow the system to operate in a mixed mode. The "follower" will be for the system-channel-less (ch. part. API)  mode, and the inactive for the "legacy" system-channel mode. 

My concern/dilema is a bit pragmatic. I am hesitant to make these two "open-heart" operations at once. I rather complete the follower work along these lines first, not touching the code paths that run in the system channel model, and only then overhaul the "system channel" paths to the same architectural pattern. There are a few reasons for that. First, caution: I don't want to break the old before the new is in place... too many changes at the same time I think. Second, as we move to the "new" composition, we might want to rethink how a "tracker" of the system channel and inactive chains work, possibly reusing some of the pieces we build for the follower.

Reply:
> My concern/dilema is a bit pragmatic. I am hesitant to make these two "open-heart" operations at once. I rather complete the follower work along these lines first, not touching the code paths that run in the system channel model, and only then overhaul the "system channel" paths to the same architectural pattern. There are a few reasons for that. First, caution: I don't want to break the old before the new is in place... too many changes at the same time I think. Second, as we move to the "new" composition, we might want to rethink how a "tracker" of the system channel and inactive chains work, possibly reusing some of the pieces we build for the follower.

I would say no need to do them "at once".  First, I would modify the 'inactive' concept to reside outside of the consensus plugin, and, once that's set and working (should be fairly easy, we've already got integration tests around the operations you'll want to test), then move on to the follower code.

I realize that this gets you to the participation API less quickly, but it means eliminating, rather than stacking debt.

To me, it comes down to a fairly simple question: Do you think having a follower, and inactive consensus plugin is the right eventual model for managing channel participation lifecycle?  If the answer is 'no', then why would we build it this way?  If the answer is 'yes', then let's discuss why.

Reply:
I would say that long term, considering only the channel participation API (no system channel), there is no need for the "inactive" concept. We only need a follower. I also agree that it does not belong to a consensus plugin, but rather at a higher level. I think we are on the same page on this. 

When I am considering the system channel mode, I think we need 3 concepts: a **chain** which runs consensus, a **tracker** which follows the last config block of every channel using the system channel, and an **on-boarding** component. When the tracker discovers a channel that the orderer was added to, it triggers the on-boarding component to complete the gap between the last block on the ledger, and the last config block that indicates membership. When on boarding is done - the consesus.Chain is started. 

Note that the latter two concepts are implemented by the InactiveChainRegistry in the cluster package. The inactive.Chain is just a place holder which does nothing. Its only use is to keep an entry in the registrar.chains map. I don't like this approach at all. Moreover, both tracking and on-boarding are implemented by a single go-routine, which is not a good idea at all.

Note that the "on-boarding" part of the story could be executed by a follower component - this is exactly what the current PR is about. So, the way I see it, getting the building blocks of the channel participation API in place will allow us to upgrade the implementation of the "system-channel" tracking & on-boarding more easily. I don't see it as stacking debt, because implementing the follower correctly in the right place will not create debt, but rather pave the way for the upgrade of the "system-channel" tracking & on-boarding.

Reply:
> I would say no need to do them "at once". First, I would modify the 'inactive' concept to reside outside of the consensus plugin, and, once that's set and working (should be fairly easy, we've already got integration tests around the operations you'll want to test), then move on to the follower code.

Just to clarify a bit more: moving the inactive.Chain out of the consensus package into a "container" or a map in the registrar is the easy part. The real complications in wiring and operations are in the InactiveChainRegistry and the go-routine that it runs.